### PR TITLE
[Duplicate] Multi-Output Classification with Machine Learning

### LINKED ISSUE #5535: [Machine Learning] Multi-Output Classification with Machine Learning
#### Proposed title of article
[Machine Learning] Multi-Output Classification with Machine Learning

#### Proposed article introduction
Multi-Output classification is a type of machine learning that predicts two different outputs at the same time. In Multi-Class machine learning, the model usually predicts only a single output but in multi-output classification, we have more than one output. Therefore, the model will be able to give more than one prediction at the same time.

For Example,  a multi-output classification model predicts the “type of fruit” and “color” of fruit at the same time.  The “type of fruit” cab be, “orange”, “mango” and “pineapple”. The “color” can be, “red”, “green”, “yellow”, and “orange". The multi-output classification can be used to solve this problem and give two prediction results.

In this tutorial, we will build a machine learning model that will be able to classify the `genre` of a movie or a series such as `drama`, `comedy`, `action`, or `thriller`.  The model will also be able to predict the `rating` of a movie or series such as `PG`, `TV14` or `TV18`. The output after prediction will be both the `genre` and `rating` of movies or series. We will use Scikit-Learn `MultiOutputClassifier` to build this model.

#### Key takeaways 
1. Multi-Class vs Multi-Label vs Multi-Output classification.
2. Techniques used to handle multi-output classification.
3.  Dataset preprocessing(Netflix dataset).
4. Building the multi-output prediction model.
5. Model predictions. (Predicts the `genre` and `rating` of movies or series)

#### Article quality
This tutorial is unique because we will compare the types of classification, Multi-Class vs Multi-Label vs Multi-Output classification. This will give us a deeper understanding of the problems before we start working on Multi-Output classification.  We will build a custom model using  Scikit-Learn `MultiOutputClassifier` algorithm. This is a practical approach and it will explain all the steps to build the model.
#### References
Please list links to any published content/research that you intend to use to support/guide this article.

### Conclusion
Finally, remove the Pre-Submission advice section and all our blockquoted notes as you fill in the form before you submit.
We look forwarding to reviewing your topic suggestion.

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by willyngashu**
## Our **Peer Reviewer and Content Moderator teams do NOT provide any revisions services.** All revisions and edits should be completed on your own forked repo (as to not take up room in the queue) in order for our team to review them all in a timely manner. 

**ANY ARTICLE SUBMITTED WITH GLARING ERRORS WILL BE IMMEDIATELY CLOSED.**

>As a rule of thumb - please be sure to only submit articles (pull requests) that are fully polished and ready to be published. Be sure to go through our resources documents for extra and 3rd party (vetted) resources to help improve overall technical writing.

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This closes #5535 

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by willyngashu** [comment]
Hello @Bayler, I have opened another PR duplicate of #5630. I had merge conflicts with the previous PR.  I have also implemented the requested changes in this new PR, kindly review.
Thanks

File: content/articles/multi-output-classification-with-machine-learning/index.md [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+from sklearn.metrics import accuracy_score
+from sklearn.multioutput import MultiOutputClassifier
+```
+Let's explain these packages that we have imported.
+
+**LogisticRegression**
+This is the algorithm used to train the model. 
+
+**CountVectorizer**
+Since we are dealing with text, we need to convert the input text into vectors of numbers. Machine learning models do not understand the raw text. The converted vectors of numbers are the representation of the original text. CountVectorizer is the most common Python package that is used to perform this process.
```
Comment (Bayler) [inline content/articles/multi-output-classification-with-machine-learning/index.md:166]:
```suggestion
Since we are dealing with text, we need to convert the input text into vectors of numbers. Machine learning models do not understand the raw text. The converted vectors of numbers are the representation of the original text. `CountVectorizer` is the most common Python package used to perform this process.
```

File: content/articles/multi-output-classification-with-machine-learning/index.md [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+
+**train_test_split**
+This is the Python package that is used for dataset splitting. In machine learning, it's essential to split a dataset into two sets. One set is to be used for training and another one for testing.
+
+**accuracy_score**
+This is used to calculate the accuracy score of the model after training.
+
+**MultiOutputClassifier**
+Since we are dealing with a multi-output classification problem, we need a more specific algorithm.
+ `MultiOutputClassifier` is the most common [Scikit-learn](https://scikit-learn.org/stable/) algorithm that is used to build this model.
```
Comment (Bayler) [inline content/articles/multi-output-classification-with-machine-learning/index.md:178]:
```suggestion
`MultiOutputClassifier` is the most common [Scikit-learn](https://scikit-learn.org/stable/) algorithm used to build this model.
```

File: content/articles/multi-output-classification-with-machine-learning/index.md [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+In the code above, we use a `test_size=0.3`. This will split our dataset so that `70%` of the dataset is used for training and `30%` will be used for testing. We have split our dataset, we are now ready to build the model.
+
+To build this model we will use the machine learning pipeline package to speed up the process of building our model. It will speed up the process by automating all the processes involved in building the model.
+
+The machine learning pipeline will automate the process of `CountVectorizer`. It will also automate the process of model training using `LogisticRegression` and `MultiOutputClassifier` algorithms.
+
+To implement this pipeline process, we will import the `Pipeline` package. 
+
+### Importing pipeline
+
```
Comment (Bayler) [inline content/articles/multi-output-classification-with-machine-learning/index.md:210]: Consider removing all the white spaces after the heading. It's unnecessary as spaces are added automatically

**Comment by willyngashu** [comment]
Hello @Bayler, I have removed the white spaces, thanks.

File: content/articles/multi-output-classification-with-machine-learning/index.md [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+In the next prediction, `TV-MA` has a higher probability of `0.46916205` compared to the other rating. That's why the model classified the rating as `TV-MA`. Using these prediction probabilities, we can see that our model was able to make the right predictions.
+
+### Conclusion
+In this tutorial, we have learned how to build a multi-output classification model. We started by cleaning our Netflix dataset to ensure that is correctly formatted before we can use it. 
+
+We then used the clean dataset to build the multi-output text classification model. We used both the `LogisticRegression` and `MultiOutputClassifier` algorithms in training the model. We implemented all the machine learning processes using the pipeline package, it sped up the process and made our work easier. 
+
+Finally, we used our model to make predictions, the trained model was able to make the right predictions. To get the multi-output classification model we have built in this tutorial, click [here](https://colab.research.google.com/drive/11-SVWysrpK0SGp7d7EYnuArlYeLsv1K0?usp=sharing).
```
Comment (Bayler) [inline content/articles/multi-output-classification-with-machine-learning/index.md:293]:
```suggestion
In this tutorial, we have learned how to build a multi-output classification model. We started by cleaning our Netflix dataset to ensure that it was correctly formatted before using it. 

We then used the clean dataset to build the multi-output text classification model. We used the `LogisticRegression` and `MultiOutputClassifier` algorithms to train the model. We implemented all the machine learning processes using the pipeline package. It sped up the process and made our work easier. 

Finally, we used our model to make predictions, and the trained model could make the right predictions. To get the multi-output classification model we have built in this tutorial, click [here](https://colab.research.google.com/drive/11-SVWysrpK0SGp7d7EYnuArlYeLsv1K0?usp=sharing).
```

File: content/articles/multi-output-classification-with-machine-learning/index.md [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+To download this Netflix dataset, click [here](https://drive.google.com/file/d/1524vMr5iCGEEJmPx2n2-TAl8EicKuxcg/view?usp=sharing)
+
+### Loading the dataset
+There are various exploratory data analysis (EDA) packages that will load our dataset. Let's import them using the following code:
+
+```python
+import pandas as pd
+import numpy as np
+```
+We will use Pandas to load the dataset. Numpy will be used to perform computational operations on our dataset. It also works well with arrays.
```
Comment (Bayler) [inline content/articles/multi-output-classification-with-machine-learning/index.md:64]: Consider rewriting the sentence, "Numpy will be used..." in active voice. For instance, "We will use Numpy...."

File: content/articles/multi-output-classification-with-machine-learning/index.md [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+
+Before we start building our model, we also need to clean our dataset. Dataset cleaning involves correctly formatting our dataset.
+
+### Text cleaning
+For text cleaning, we will convert all our text data into lower case and also remove stop words. We will use the NeatText Python package to perform this process. We will install Neattext using the following code:
+
+```bash
+! pip install neattext
+```
+Let's import the Neattext functions that will be used for text cleaning.
```
Comment (Bayler) [inline content/articles/multi-output-classification-with-machine-learning/index.md:125]: This is another case of using passive voice. Instead of using "will be used...", consider replacing that with "...will use"

File: content/articles/multi-output-classification-with-machine-learning/index.md [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+from sklearn.linear_model import LogisticRegression
+from sklearn.feature_extraction.text import CountVectorizer
+from sklearn.model_selection import train_test_split
+from sklearn.metrics import accuracy_score
+from sklearn.multioutput import MultiOutputClassifier
+```
+Let's explain these packages that we have imported.
+
+**LogisticRegression**
+This is the algorithm used to train the model.
```
Comment (Bayler) [inline content/articles/multi-output-classification-with-machine-learning/index.md:158]: Consider using "It" instead of "This"

File: content/articles/multi-output-classification-with-machine-learning/index.md [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+**CountVectorizer**
+Since we are dealing with text, we need to convert the input text into vectors of numbers. Machine learning models do not understand the raw text. The converted vectors of numbers are the representation of the original text. `CountVectorizer` is the most common Python package used to perform this process.
+
+For further reading on CountVectorizer and how they convert raw text into vectors of numbers, click [here](https://towardsdatascience.com/basics-of-countvectorizer-e26677900f9c)
+
+**train_test_split**
+This is the Python package that is used for dataset splitting. In machine learning, it's essential to split a dataset into two sets. One set is to be used for training and another one for testing.
+
+**accuracy_score**
+This is used to calculate the accuracy score of the model after training.
```
Comment (Bayler) [inline content/articles/multi-output-classification-with-machine-learning/index.md:169]: Consider using "It" instead of "This" in this section

File: content/articles/multi-output-classification-with-machine-learning/index.md [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+
+The next step is to split our dataset using the `train_test_split` method.
+
+### Dataset splitting
+To split the dataset into two, use this code:
+
+```python
+x_train,x_test,y_train,y_test = train_test_split(Xfeatures,ylabels,test_size=0.3,random_state=7)
+```
+In the code above, we use a `test_size=0.3`. This will split our dataset so that `70%` of the dataset is used for training and `30%` will be used for testing. We have split our dataset, we are now ready to build the model.
```
Comment (Bayler) [inline content/articles/multi-output-classification-with-machine-learning/index.md:194]:
```suggestion
In the code above, we use a `test_size=0.3`. It will split our dataset so that `70%` of the dataset is used for training and `30%` for testing. We have split our dataset, and we are now ready to build the model.
```

File: content/articles/multi-output-classification-with-machine-learning/index.md [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+
+To build this model we will use the machine learning pipeline package to speed up the process of building our model. It will speed up the process by automating all the processes involved in building the model.
+
+The machine learning pipeline will automate the process of `CountVectorizer`. It will also automate the process of model training using `LogisticRegression` and `MultiOutputClassifier` algorithms.
+
+To implement this pipeline process, we will import the `Pipeline` package.
```
Comment (Bayler) [inline content/articles/multi-output-classification-with-machine-learning/index.md:200]:
```suggestion
We will import the `Pipeline` package to implement this pipeline process. 
```

File: content/articles/multi-output-classification-with-machine-learning/index.md [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+```python
+pipe_lr = Pipeline(steps=[('cv',CountVectorizer()),
+                          ('lr_multi',MultiOutputClassifier(LogisticRegression()))])
+```
+Now that we have initialized the processes, let's fit the pipeline into our training dataset. This will enable the model to learn from the dataset. To fit the pipeline, use the following code: 
+
+```python
+pipe_lr.fit(x_train,y_train)
+```
+During this stage, the two processes will run automatically and produce a trained model as shown below.
```
Comment (Bayler) [inline content/articles/multi-output-classification-with-machine-learning/index.md:226]:
```suggestion
The two processes will run automatically during this stage and produce a trained model, as shown below:
```

File: content/articles/multi-output-classification-with-machine-learning/index.md [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+
+```python
+pipe_lr.score(x_test,y_test)
+```
+The accuracy score is shown below:
+
+```bash
+0.8969221004536385
+```
+The accuracy score for our model is `0.896922`, this represents `89.6922%`. This is a good accuracy score, we can use this trained model to make predictions.
```
Comment (Bayler) [inline content/articles/multi-output-classification-with-machine-learning/index.md:240]:
```suggestion
The accuracy score for our model is `0.896922`. This represents `89.6922%`. It is a good accuracy score, and we can use this trained model to make predictions.
```

File: content/articles/multi-output-classification-with-machine-learning/index.md [hunk 13]
```diff
@@ -0,0 +1,303 @@
...
+
+### Making predictions
+To make a prediction we need to extract a sample input text. To extract a sample text run this code:
```
Comment (Bayler) [inline content/articles/multi-output-classification-with-machine-learning/index.md:243]: Please add commas in these two sentences where necessary

File: content/articles/multi-output-classification-with-machine-learning/index.md [hunk 14]
```diff
@@ -0,0 +1,303 @@
...
+
+```python
+print(x_test.iloc[0])
+```
+The output of this sample text is `the midnight sky`. Let's save this text in a variable.
+
+```python
+pred1 = x_test.iloc[0]
+```
+The model will use this input text to make a prediction. The model should be able to classify the input text as either a `Movie` or `TV Show` and also provide its rating.
```
Comment (Bayler) [inline content/articles/multi-output-classification-with-machine-learning/index.md:253]:
```suggestion
The model will use this input text to make a prediction. The model should classify the input text as either a `Movie` or `TV Show` and provide its rating.
```

File: content/articles/multi-output-classification-with-machine-learning/index.md [hunk 15]
```diff
@@ -0,0 +1,303 @@
...
+```
+The output is shown below:
+
+```bash
+[array(['Movie', 'TV Show'], dtype=object), array(['PG-13', 'R', 'TV-14', 'TV-MA', 'TV-PG', 'TV-Y'], dtype=object)]
+[array([[0.74445483, 0.25554517]]),
+ array([[0.12310188, 0.07038494, 0.21476461, 0.46916205, 0.10270243,
+         0.01988409]])]
+```
+From the output above we can see `Movie` had a higher probability of `0.74445483` compared to `0.25554517` of `TV Show`. That's why the model classified the text as `Movie`.
```
Comment (Bayler) [inline content/articles/multi-output-classification-with-machine-learning/index.md:284]:
```suggestion
From the output above, we can see `Movie` had a higher probability of `0.74445483` than `0.25554517` of `TV Show`. That's why the model classified the text as `Movie`.
```

**Comment by Bayler** [comment]
@willyngashu Please work on the requested revisions before we can move on for final review.

**Comment by willyngashu** [comment]
@Bayler, I have worked on the requested revisions

**Comment by Bayler** [comment]
@willyngashu Please take note of all the commits I've made in this tutorial so as to better your next tutorial.

**Review by louisefindlay23 — APPROVED** [review]
Final review complete
